### PR TITLE
rgw_sal_motr : [CORTX-34167] Modify error code and error message of GET/HEAD object API

### PR DESCRIPTION
**Pull request by priyanka25081999**
Problem Statement:
GET/HEAD Object API with version id of a delete marker gives `404 (NoSuchKey)` error in the response.

Solution: 
Modified the error code and error message of GET/HEAD object API response to `405 (MethodNotAllowed)` in case of delete-marker.

Signed-off-by: Priyanka Salunke <priyanka.s.salunke@seagate.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by priyanka25081999** [comment]
1N k8s auto-triggered deployment is successful - https://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Cortx-Deployment/job/Generic/job/cortx-rgw/208/ 
image

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -1933,8 +1933,19 @@ int MotrObject::MotrReadOp::prepare(optional_yield y, const DoutPrefixProvider*
...
+  // In GET/HEAD object API, return "MethodNotAllowed"
+  // if delete-marker is the latest object entry
+  // Else, return "NoSuchKey" error
+  if (ent.is_delete_marker()) {
+    if (source->get_instance() == ent.key.instance && ent.key.instance != "") {
+      s->err.message = "The specified method is not allowed against this resource.";
+      return -ERR_METHOD_NOT_ALLOWED;
+    }
+    else
+      return -ENOENT;
```
Comment (jjxsg) [inline src/rgw/rgw_sal_motr.cc:1947]: You don't need `else` here
```suggestion
    return -ENOENT;
```

File: src/rgw/rgw_sal_motr.cc [hunk 2]
```diff
@@ -1933,8 +1933,19 @@ int MotrObject::MotrReadOp::prepare(optional_yield y, const DoutPrefixProvider*
...
-  if (ent.is_delete_marker())
-    return -ENOENT;
+  req_state* s = static_cast<req_state*>(rctx->get_private());
+
+  // In GET/HEAD object API, return "MethodNotAllowed"
+  // if delete-marker is the latest object entry
+  // Else, return "NoSuchKey" error
+  if (ent.is_delete_marker()) {
+    if (source->get_instance() == ent.key.instance && ent.key.instance != "") {
+      s->err.message = "The specified method is not allowed against this resource.";
```
Comment (jjxsg) [inline src/rgw/rgw_sal_motr.cc:1943]: I believe a debug log entry stating why `MethodNotAllowed` error occurred would be helpful.

Comment (cdeshmukh) [reply src/rgw/rgw_sal_motr.cc:1943]: agree. 
infact error message also needs to be specific, something like
"   s->err.message = "The <GET/HEAD whatever applies from request> method not allowed against delete marker <version-id>";"

File: src/rgw/rgw_sal_motr.cc (hunk 1 above)
Comment (cdeshmukh) [reply src/rgw/rgw_sal_motr.cc:1947]: > You don't need `else` here

else is needed still, to keep ENOENT response for other regular cases. Here code diff is not reflecting properly as i see, here is the screenshot changed code
image

Comment (jjxsg) [reply src/rgw/rgw_sal_motr.cc:1947]: I meant, `return -ENOENT;` is required but no need to keep it inside `else`. 

These two code snippets will give the same results:
```
{
  if (something) {
    return this;
  }
  else
    return that;
}
```
```
{
  if (something) {
    return this;
  }
  return that;
}
```

Comment (priyanka25081999) [reply src/rgw/rgw_sal_motr.cc]: Done. Followed 2nd approach.

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (priyanka25081999) [reply src/rgw/rgw_sal_motr.cc:1945]: Yes, added the debug log.

Comment (priyanka25081999) [reply src/rgw/rgw_sal_motr.cc:1945]:
```
`infact error message also needs to be specific, something like
" s->err.message = "The <GET/HEAD whatever applies from request> method not allowed against delete marker ";"`
```
Currently, MGW returns "Unknown" in the error message. To change it and make it inline with AWS, I have added the "The specified method is not allowed against this resource." error message. Now MGW is inline with AWS.

**Review by cdeshmukh — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc [hunk 3]
```diff
@@ -1933,8 +1933,20 @@ int MotrObject::MotrReadOp::prepare(optional_yield y, const DoutPrefixProvider*
...
-  if (ent.is_delete_marker())
+  req_state* s = static_cast<req_state*>(rctx->get_private());
+
+  // In GET/HEAD object API, return "MethodNotAllowed"
+  // if delete-marker is the latest object entry
+  // Else, return "NoSuchKey" error
+  if (ent.is_delete_marker()) {
+    if (source->get_instance() == ent.key.instance && ent.key.instance != "") {
+      ldpp_dout(dpp, LOG_DEBUG) <<__func__ << "The GET/HEAD object with version-id of "
+                                            "delete-marker is not allowed." << dendl;
```
Comment (jjxsg) [inline src/rgw/rgw_sal_motr.cc:1944]:
```suggestion
      ldpp_dout(dpp, LOG_DEBUG) <<__func__ << ": GET/HEAD object with version-id of "
                                            "delete-marker is not allowed." << dendl;
```

**Review by jjxsg — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc (hunk 3 above)
Comment (priyanka25081999) [reply src/rgw/rgw_sal_motr.cc]: Yeah, corrected the statement now. Thanks!
Also, added DEBUG word in log statement
```
ldpp_dout(dpp, LOG_DEBUG) <<__func__ << ": DEBUG: The GET/HEAD object with version-id of "
                                            "delete-marker is not allowed." << dendl;
```

**Review by shraddhaghatol — APPROVED** [review]

**Comment by priyanka25081999** [comment]
Sanity is successful - https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/S3Server/job/S3-Compatibility-Test-Ceph-RGW/237/ [PASSED]
image

**Pull request merged**